src/init.cpp

```diff
@@ -125,16 +125,25 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
 
 std::atomic<bool> fRequestShutdown(false);
 std::atomic<bool> fDumpMempoolLater(false);
+static std::mutex mutexShutdownWait;
+static std::condition_variable cvShutdownWait;
 
 void StartShutdown()
 {
     fRequestShutdown = true;
+    std::unique_lock<std::mutex> lockShutdownWait(mutexShutdownWait);
+    cvShutdownWait.notify_all();
```

Comment:
Are you sure this is allowed in signal handlers? From what I remember that is the reason for the circuitous polling. If not we need a special path for signals and one for the rest (like AbortNode).